doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+typedef struct ElektraKeyset ElektraKeyset;
+
+// 0 always indicates success, other values depend on function
+typedef int ElektraErrorCode;
+
+// Note: breaking change compared to current values
+typedef enum {
+    ELEKTRA_KEY_CP_NAME   = 1<<0,
+    ELEKTRA_KEY_CP_VALUE  = 1<<1,
+    ELEKTRA_KEY_CP_META   = 1<<2,
+    ELEKTRA_KEY_CP_ALL    = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+} ElektraKeyCopyFlag;
+
+void elektraKeynamePushPart (ElektraKeyname * name, const char * part);
+void elektraKeynamePopPart (ElektraKeyname * name);
+
+ElektraKey * elektraKeyNew (const ElektraKeyname * name);
+ElektraKey * elektraKeyRetain (const ElektraKey * key);
+void elektraKeyRelease (const ElektraKey * key);
+
+const ElektraKeyname * elektraKeyGetName (const ElektraKey * key);
+// Note: creates copy of name
+ElektraErrorCode elektraKeySetName (ElektraKey * key, const ElektraKeyname * name);
+// increments nameLock
+void elektraKeyLockName (ElektraKey * key);
+// decrements nameLock
+void elektraKeyUnlockName (ElektraKey * key);
+bool elektraKeyIsNameLocked (const ElektraKey * key);
+
+const ElektraKeyvalue * elektraKeyGetValue (const ElektraKey * key);
+// Note: creates copy of value
+ElektraErrorCode elektraKeySetValue (ElektraKey * key, const ElektraKeyvalue * value);
+
+/**
+ * @returns a `const ElektraKeyset *` that can safely be cast to `ElektraKeyset *`, iff a non-const `ElektraKey *` was passed as @p key
+ */
+ELEKTRA_CONST ElektraKeyset * elektraKeyGetMeta (const ElektraKey * key);
+
+ElektraErrorCode elektraKeyIsBelow (const ElektraKey * first, const ElektraKey * second);
+ElektraErrorCode elektraKeyCompare (const ElektraKey * first, const ElektraKey * second);
+
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * src, ElektraKeyCopyFlag flags);
+static inline ElektraKey * elektraKeyDup (ElektraKey * key, ElektraKeyCopyFlag flags)
+{
+    ElektraKey * dest = flags & ELEKTRA_KEY_CP_NAME
+        ? elektraKeyNew (elektraKeyGetName (key))
+        : elektraKeyNew (&(ElektraKeyname){.ns = ELEKTRA_NS_CASCADING, .name = "", .size = 0});
+    return elektraKeyCopy (dest, key, (ElektraKeyCopyFlag)(flags & ~ELEKTRA_KEY_CP_NAME));
+}
+
+
+ElektraKeyset * elektraKeysetNew (size_t prealloc);
+ElektraKeyset * elektraKeysetRetain (const ElektraKeyset * ks);
+void elektraKeysetRelease (const ElektraKeyset * ks);
+
+size_t elektraKeysetSize (ElektraKeyset * ks);
+
+ElektraErrorCode elektraKeysetInsert (ElektraKeyset * ks, ElektraKey * key);
+ElektraErrorCode elektraKeysetInsertAll (ElektraKeyset * ks, ElektraKeyset * other);
+static inline ElektraErrorCode elektraKeysetInsertAndRelease (ElektraKeyset * ks, ElektraKey * key)
+{
+    ElektraErrorCode error = elektraKeysetInsert (ks, key);
+    elektraKeyRelease (key);
+    return error;
+}
+
+ElektraKey * elektraKeysetGet (ElektraKeyset * ks, size_t index);
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
```

Comment:
This would need some use case.

Reply:
See #4797. It is the basis for the cut functions described below.